rasa/core/lock.py

```diff
@@ -0,0 +1,154 @@
+import json
+import logging
+from collections import deque
+from typing import Text, Optional, Union, Deque, Dict, Any
+
+import time
+
+logger = logging.getLogger(__name__)
+
+NO_TICKET_ISSUED = -1  # index of latest issued ticket if no tickets exist
+
+
+class Ticket:
+    def __init__(self, number: int, expires: float) -> None:
+        self.number = number
+        self.expires = expires
+
+    def has_expired(self) -> bool:
+        return time.time() > self.expires
+
+    def as_dict(self) -> Dict[Text, Any]:
+        return dict(number=self.number, expires=self.expires)
+
+    def dumps(self) -> Text:
+        """Return json dump of `Ticket` as dictionary."""
+
+        return json.dumps(self.as_dict())
+
+    @classmethod
+    def from_dict(cls, data: Dict[Text, Union[int, float]]) -> "Ticket":
+        """Creates `Ticket` from dictionary."""
+
+        return cls(number=data["number"], expires=data["expires"])
+
+    def __repr__(self) -> Text:
+        return "Ticket(number: {}, expires: {})".format(self.number, self.expires)
+
+
+class TicketLock:
+    """Locking mechanism that issues tickets managing access to conversation IDs.
+
+    Tickets are issued in the order in which they are requested. A detailed
+    explanation of the ticket lock algorithm can be found at
+    http://pages.cs.wisc.edu/~remzi/OSTEP/threads-locks.pdf#page=13
+    """
+
+    def __init__(
+        self, conversation_id: Text, tickets: Optional[Deque[Ticket]] = None
```

Comment:
why do we actually need to store the whole queue and not only the last issued number?

Reply:
yes, we need to keep track of which tickets have expired, also the lock maintains a count of ticket numbers which are tied to the index in the queue